client/network/src/persist_peers.rs

```diff
@@ -0,0 +1,308 @@
+	}
+
+	pub fn report_peer_addr(
+		&mut self,
+		peer_id: &PeerId,
+		protocol: impl AsRef<[u8]>,
+		addr: &Multiaddr,
+	) {
+		let protocol = String::from_utf8(protocol.as_ref().to_owned()).expect(
+			"According to the `crate::discovery::protocol_name_from_protocol_id` \
+					and `<ProtocolId as AsRef<str>>` it's a correct UTF-8 string",
+		);
+
+		let entries = self
+			.protocols
+			.entry(protocol)
+			.or_insert_with(|| LruCache::new(PEER_ADDRS_CACHE_SIZE));
+		if let Some(peer_addrs) = entries.get_mut(peer_id) {
+			peer_addrs.insert(addr.to_owned());
+		} else {
+			entries.push(peer_id.to_owned(), [addr.to_owned()].into_iter().collect());
+		}
+	}
+
+	pub fn peer_addrs<'a>(
+		&'a mut self,
+		peer_id: &'a PeerId,
+		protocols: impl IntoIterator<Item = impl AsRef<[u8]> + 'a>,
+	) -> impl Iterator<Item = &'a Multiaddr> {
+		let protocols = protocols.into_iter().collect::<Vec<_>>();
+
+		self.protocols
+			.iter_mut()
+			.filter_map(move |(protocol, entries)| {
+				if protocols.iter().any(|p| p.as_ref() == protocol.as_bytes()) {
+					Some(entries)
+				} else {
+					None
+				}
+			})
+			.flat_map(|entries| entries.get(peer_id).into_iter())
+			.flat_map(IntoIterator::into_iter)
+	}
+
+	pub fn poll(&mut self, cx: &mut Context) -> Poll<Never> {
+		if let Some(busy_future) = self.busy.as_mut() {
+			if let Poll::Ready(result) = busy_future.poll_unpin(cx) {
+				self.busy = None;
+				self.flushed_at = Instant::now();
+
+				if let Err(reason) = result {
+					log::warn!("Failed to persist peer addresses: {}", reason);
+				}
+			}
+		} else if self.flushed_at.elapsed() > FLUSH_INTERVAL {
+			let entries = self
+				.protocols
+				.iter()
+				.map(|(protocol, entries)| {
+					let entries = entries
+						.iter()
+						.map(|(peer_id, addrs)| {
+							let peer_id = peer_id.to_base58();
+							let addrs = addrs.into_iter().cloned().collect();
+
+							persist_peer_addrs::PeerEntry { peer_id, addrs }
+						})
+						.collect::<Vec<_>>();
+					(protocol.to_owned(), entries)
+				})
+				.collect();
+
+			let busy_future = persist_peer_addrs::persist(Arc::clone(&self.paths), entries).boxed();
+			self.busy = Some(busy_future);
+		}
+		Poll::Pending
+	}
+}
+
+mod persist_peer_addrs {
+	use super::*;
+	use tokio::io::AsyncWriteExt;
+
+	#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+	pub(super) struct PeerEntry {
+		pub peer_id: String,
+		pub addrs: Vec<Multiaddr>,
+	}
+
+	pub(super) async fn persist(
+		paths: Arc<Paths>,
+		protocols: HashMap<ProtocolType, Vec<PeerEntry>>,
+	) -> Result<(), io::Error> {
+		let mut tmp_file = tokio::fs::OpenOptions::new()
+			.create(true)
+			.write(true)
+			.open(&paths.tmp_path)
+			.await?;
+		let serialized = serde_json::to_vec_pretty(&protocols)?;
+
+		tmp_file.write_all(&serialized).await?;
+		tmp_file.flush().await?;
+		std::mem::drop(tmp_file);
+
+		tokio::fs::rename(&paths.tmp_path, &paths.path).await?;
```

Comment:
You're doing this whole dance twice; I'd suggest to move it to a helper method, e.g.:

```
async fn serialize_json_into_file_atomic(path: impl AsRef<Path>, payload: impl Serialize) -> Result<(), std::io::Error>
```

or something like this. Same for loading.

Reply:
Moved that shared routines into `sc-network::utils::json_file::{save, load}`.